File: src/test/resources/org/bau/benchmarks/fannkuch.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
use std::env;

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 {
        args[1].parse().unwrap()
    } else {
        12
    };
    println!("Pfannkuchen({}) = {}", n, fannkuch(n));
}

fn fannkuch(n: usize) -> i32 {
    let mut perm1 = vec![0; n];
    for i in 0..n {
        perm1[i] = i;
    }
    let mut perm = vec![0; n];
    let mut count = vec![0; n];
    let mut flips = 0;
    let mut nperm = 0;
    let mut checksum = 0;
    let mut r = n;
    while r > 0 {
        while r > 1 {
            count[r - 1] = r;
            r -= 1;
        }
        for i in 0..n {
            perm[i] = perm1[i];
        }
        // Count flips
        let mut f = 0;
        let mut k = perm[0];
        while k != 0 {
            let mut i = 0;
            while 2 * i < k {
                perm.swap(i, k - i);
                i += 1;
            }
            k = perm[0];
            f += 1;
        }
        if f > flips {
            flips = f;
        }
        if nperm & 1 == 0 {
            checksum += f;
        } else {
            checksum -= f;
        }
        // Generate next permutation
        let mut more = true;
        while more {
            if r == n {
                println!("{}", checksum);
                return flips;
            }
            let p0 = perm1[0];
            let mut i = 0;
            while i < r {
                perm1[i] = perm1[i + 1];
                i += 1;
            }
            perm1[r] = p0;
            count[r] -= 1;
            if count[r] > 0 {
                more = false;
            } else {
                r += 1;
            }
        }
        nperm += 1;
    }
    flips
}

File: src/test/resources/org/bau/benchmarks/fannkuch.go
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

package main

import (
    "flag"
    "fmt"
    "strconv"
)

func main() {
    flag.Parse()
    n := 4
    if flag.NArg() > 0 { n, _ = strconv.Atoi(flag.Arg(0)) }
    fmt.Printf("Pfannkuchen(%d) = %d\n", n, fannkuch(n))
}

func fannkuch(n int) int {
    perm1 := make([]int, n)
    for i := 0; i < n; i++ {
        perm1[i] = i
    }
    perm := make([]int, n)
    count := make([]int, n)
    f, k, r, flips, nperm, checksum := 0, 0, 0, 0, 0, 0
    r = n
    for r > 0 {
        for r > 1 {
            count[r - 1] = r
            r--
        }
        for i := 0; i < n; i++ {
            perm[i] = perm1[i]
        }
        // Count flips and update max and checksum
        f = 0
        k = perm[0]
        for k != 0 {
            for i := 0; 2 * i < k; i++ {
                t := perm[i]
                perm[i] = perm[k - i]
                perm[k - i] = t
            }
            k = perm[0]
            f++
        }
        if f > flips {
            flips = f
        }
        if (nperm & 1) == 0 {
            checksum += f
        } else {
            checksum -= f
        }
        // Use incremental change to generate another permutation
        for {
             if r == n {
                fmt.Println(checksum)
                return flips
             }
             p0 := perm1[0]
             for i := 0; i < r; {
                j := i + 1
                perm1[i] = perm1[j]
                i = j
            }
            perm1[r] = p0
            count[r]--
            if count[r] > 0 {
                break;
            }
            r++
        }
        nperm++
    }
    return flips
}

